packages/wallet/src/redux/selectors.ts

```diff
@@ -0,0 +1,24 @@
+import { OpenedState, OPENING } from './channel-state/state';
+import * as walletStates from './state';
+
+export const getOpenedChannelState = (
+  state: walletStates.Initialized,
+  channelId: string,
+): OpenedState => {
+  const channelStatus = state.channelState.initializedChannels[channelId];
+  if (!channelStatus) {
+    throw new Error(`Could not find any initialized channel state for channel ${channelId}.`);
+  }
+  if (channelStatus.stage === OPENING) {
+    throw new Error(`Channel ${channelId} is still in the process of being opened.`);
+  }
+  return channelStatus;
+};
+
+// TODO: Ideally we should be able to pass in a expected state type and have the selector either return that type of state or throw an error
+export const getIndirectFundingState = (state: walletStates.Initialized) => {
```

Comment:
This can, and probably should, be typed to return a state that is known to have an `indirectFunding` property.

Reply:
You would need to define an interface, like
```
interface HasIndirectFunding extends Initialized {
  indirectFunding: IndirectFunding
}
```

(there is probably some typescript generic that does this?)

Reply:
This returns the `IndirectFundingState` as opposed to the `walletState` so I don't think we need that interface. The return type is just `IndirectFundingState`.

Ideally it'd be nice to do this: `const indirectFundingState = getIndirectFundingState(walletState, internalStates.WAIT_FOR_PRE_FUND_SETUP1)` and have `indirectFundingState` be typed as `WaitForPrefundSetup1` for you (or throw an error if the type doesn't match).

Reply:
>This returns the IndirectFundingState as opposed to the walletState so I don't think we need that interface. The return type is just IndirectFundingState.

Right, of course.

>Ideally it'd be nice to do this: const indirectFundingState = getIndirectFundingState(walletState, internalStates.WAIT_FOR_PRE_FUND_SETUP1) and have indirectFundingState be typed as WaitForPrefundSetup1 for you (or throw an error if the type doesn't match).

That signature suggests, to me, that you're looking for _the_ `indirectFunding` "thing" (possibly among other `indirectFunding` "things") that is a ` internalStates.WAIT_FOR_PRE_FUND_SETUP1`. It seems wrong for a function's purpose to be
> get me the indirect funding data, and please check that it is this type

What if you wanted to get the indirect funding and didn't care what the type was?

Reply:
I think you're right that we want a selector that just gets the `indirectFundingState`. However a lot of cases where we want to fetch the `indirectFundingState` and we expect it to be a certain type.

It'd be nice to have a `selector` or helper function that gets `indirectFundingState`, checks its the correct type (throws an error if not) and returns the state correctly typed. So instead of doing this and possibly missing the case where the state is not what we expect ``` 
 const indirectFundingState = selectors.getIndirectFundingState(
    state,
  ) as states.WaitForPostFundSetup1;
```

we could do this 
``` cont waitForPostFund = selectors.getTypedIndirectFundingState(state, states.WAIT_FOR_POST_SETUP_1.);
// Since the state is typed as WaitForPostFundSetup1 we can access the ledgerId property.
const ledgerID = waitForPostFund.ledgerId 
```